org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ParameterizedTypeBinding.java

```diff
@@ -1108,26 +1110,71 @@ public boolean isRawSubstitution() {
 
 	@Override
 	public ReferenceBinding[] permittedTypes() {
-		ReferenceBinding[] permTypes = this.type.permittedTypes();
-		List<ReferenceBinding> applicablePermTypes = new ArrayList<>();
-		for (ReferenceBinding pt : permTypes) {
-			ReferenceBinding permittedTypeAvatar = pt;
-			if (pt.isRawType()) {
-				ReferenceBinding ptRef = pt.actualType();
-				ReferenceBinding enclosingType1 = ptRef.enclosingType();
-				if (enclosingType1 != null) {
-					// don't use TypeSystem.getParameterizedType below as this is just for temporary check.
-					ParameterizedTypeBinding ptb = new ParameterizedTypeBinding(ptRef, this.arguments, ptRef.enclosingType(), this.environment);
-					ptb.superclass();
-					ptb.superInterfaces();
-					permittedTypeAvatar = ptb;
+		List<ReferenceBinding> permittedTypes = new ArrayList<>();
+NextPermittedType:
+		for (ReferenceBinding pt : this.type.permittedTypes()) {
+			// Step 1: Gather all type variables that would need to be solved.
+			Map<TypeVariableBinding, TypeBinding> map = new HashMap<>();
+			TypeBinding current = pt;
+			do {
+				if (current.kind() == Binding.GENERIC_TYPE) {
+					for (TypeVariableBinding tvb : current.typeVariables()) {
+						map.put(tvb,  null);
```

Comment:
> How would anyone observe the effect of this put()?

Isn't it more pertinent to ask "How would anyone observe the effect of this do while loop?" :) 

So yes it would seem the entire enclosing `do {} while` could be gotten rid of - which I will do along with anymore follow up actions on behalf of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3038

Reply:
> How would anyone observe the effect of this put()?

PR https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3149 addresses this review comment, thanks!